quic/s2n-quic-platform/src/socket/io/tx.rs

```diff
@@ -0,0 +1,439 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{features::Gso, message::Message, socket::ring::Producer};
+use core::task::{Context, Poll};
+use s2n_quic_core::{
+    event,
+    inet::ExplicitCongestionNotification,
+    io::tx,
+    path::{Handle as _, MaxMtu},
+};
+
+/// Structure for sending messages to producer channels
+pub struct Tx<T: Message> {
+    channels: Vec<Producer<T>>,
+    gso: Gso,
+    max_mtu: usize,
+    is_full: bool,
+}
+
+impl<T: Message> Tx<T> {
+    #[inline]
+    pub fn new(channels: Vec<Producer<T>>, gso: Gso, max_mtu: MaxMtu) -> Self {
+        Self {
+            channels,
+            gso,
+            max_mtu: max_mtu.into(),
+            is_full: true,
+        }
+    }
+}
+
+impl<T: Message> tx::Tx for Tx<T> {
+    type PathHandle = T::Handle;
+    type Queue = TxQueue<'static, T>;
+    type Error = ();
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context) -> Poll<Result<(), Self::Error>> {
+        // We only need to poll for capacity if we completely filled up all of the channels.
```

Comment:
Right - my thinking is that this is right now making an assumption that sending to the 2nd queue is cheap enough that a wait on the 1st queue is always more expensive. I could imagine instead e.g. spinning on the 1st queue for 1-2 cycles and then switching to sending into the 2nd queue (and so on). In practice I doubt this makes much sense...

I do think it would be good at some point to wire up metrics of some kind so we can report on whether we're using 1 queue or constantly round-robining.

Reply:
Yeah we definitely have some assumptions we need to validate here. Currently, there isn't even an option to do multi-socket queues so we can revisit strategies once we expose that :)